Feature: Enhance User Creation Console Command

ardelato opened the pull request:
## Description

This updates the `user:create` command to include options for role assignment and consent dates. The `docker_run.sh` script has been modified to utilize the new command format for creating an admin user with auto-consent.

This will also be useful for creating admin users during the initialization of a new instance of the app.

qa_req 0

ardelato committed:
feat(user): enhance user creation command with role and consent options

Updated the user:create command to include options for role assignment and
consent dates. The docker_run.sh script has been modified to utilize the
new command format for creating an admin user with auto-consent.

This will also be useful for creating admin users during the initialization
of a new instance of the app.

mlahargou commented on app/Console/Commands/UserCreate.php:
```diff
@@ -18,14 +19,19 @@ class UserCreate extends Command
-    protected $signature = 'user:create {name} {email} {password} {language} {repair_network_id}';
+    protected $signature = 'user:create {name} {email} {password} {language} {repair_network_id} 
+                           {--role=4 : Role ID (1=ROOT, 2=ADMINISTRATOR, 3=HOST, 4=RESTARTER, 5=GUEST, 6=NETWORK_COORDINATOR)}
+                           {--consent-past-data= : Consent date for past data in YYYY-MM-DD format, defaults to today}
```
Is this comment correct? I don't think it defaults to today unless `--auto-consent` is set, right?

(Same with the other args)

ardelato replied:
Addressed in [c4a6899](https://github.com/iFixit/restarters/pull/15/commits/c4a68992bb961b27f2857873e7d91ea41389338d)

mlahargou commented on app/Console/Commands/UserCreate.php:
```diff
@@ -71,23 +87,52 @@ public function handle(DiscourseService $discourseService): void
+                'username' => '',
+                'wiki_sync_status' => WikiSyncStatus::CreateAtLogin,
+                'language' => $language,
+                'repair_network' => $repair_network_id,
+            ];
+
+            // Add consent data if provided
+            if (!empty($consent_past_data)) {
+                $userData['consent_past_data'] = $consent_past_data;
+            }
+            
+            if (!empty($consent_future_data)) {
+                $userData['consent_future_data'] = $consent_future_data;
+            }
+            
+            if (!empty($consent_gdpr)) {
+                $userData['consent_gdpr'] = $consent_gdpr;
+            }
```
These values are either string dates or null, right? I think we can simplify it.

```suggestion
                'repair_network' => $repair_network_id,
                'consent_past_data' => $consent_past_data ?? '',
                'consent_future_data' => $consent_future_data ?? '',
                'consent_gdpr' => $consent_gdpr ?? '',
            ];
```

ardelato replied:
Addressed in [c428255](https://github.com/iFixit/restarters/pull/15/commits/c4282554dc2afebfc3aa1dab2415c40d36480569)

mlahargou commented on app/Console/Commands/UserCreate.php:
```diff
@@ -71,23 +87,52 @@ public function handle(DiscourseService $discourseService): void
                 $this->info("User created #" . $user->id);
+                $this->info("Role: " . $this->getRoleName($user->role));
+                
+                if (!empty($userData['consent_past_data']) || !empty($userData['consent_future_data']) || !empty($userData['consent_gdpr'])) {
+                    $this->info("Consent dates set:");
+                    if (!empty($userData['consent_past_data'])) {
+                        $this->info("- Past data: {$userData['consent_past_data']}");
+                    }
+                    if (!empty($userData['consent_future_data'])) {
+                        $this->info("- Future data: {$userData['consent_future_data']}");
+                    }
+                    if (!empty($userData['consent_gdpr'])) {
+                        $this->info("- GDPR: {$userData['consent_gdpr']}");
+                    }
+                }
```
```suggestion
                if ($userData['consent_past_data'] || $userData['consent_future_data'] || $userData['consent_gdpr']) {
                    $this->info("Consent dates set:");
                    if ($userData['consent_past_data']) {
                        $this->info("- Past data: {$userData['consent_past_data']}");
                    }
                    if ($userData['consent_future_data']) {
                        $this->info("- Future data: {$userData['consent_future_data']}");
                    }
                    if ($userData['consent_gdpr']) {
                        $this->info("- GDPR: {$userData['consent_gdpr']}");
                    }
                }
```

ardelato replied:
Addressed in [c428255](https://github.com/iFixit/restarters/pull/15/commits/c4282554dc2afebfc3aa1dab2415c40d36480569)

I ended up just dropping the logs. They were more beneficial in the beginning but now not anymore.

mlahargou reviewed (commented):
dev_block 📱 on a couple suggestions. I'm a little confused what's supposed to happen if the data args aren't set. Should it be left off of the user data, or an empty string, or today's date?

ardelato committed:
refactor(user): remove auto-consent option from user:create command

The auto-consent option was a bit redundant given the descriptions of
the other consent options. As such, I removed it and included a validator
for the consent dates.

ardelato committed:
refactor(user): remove redundant checks for consent data

Removed redundant checks for consent data in the user:create command,
since the consent data will never be empty. In addition, we don't really
need to print the consent data to the console.

ardelato commented:
un_dev_block ⚡

mlahargou commented on app/Console/Commands/UserCreate.php:
```diff
@@ -101,4 +133,33 @@ public function handle(DiscourseService $discourseService): void
+    }
+
+    /**
+     * Validate a date string against a format.
+     */
+    private function validateDate($date, $format = 'Y-m-d')
+    {
+        $d = \DateTime::createFromFormat($format, $date);
```
[super minor] We never pass in a `format`, so it doesn't really need to be a param.

```suggestion
     * Validate a date string for proper format.
     */
    private function validateDate($date)
    {
        $format = 'Y-m-d';
        $d = \DateTime::createFromFormat($format, $date);
```

ardelato replied:
Addressed in [f371896](https://github.com/iFixit/restarters/pull/15/commits/f371896246dac0eac6853fee70b487abcaaf1bab)

mlahargou reviewed (approved):
CR 📱

ardelato committed:
refactor(user): simplify date validation in user:create command

Updated the validateDate method to remove the redundant format parameter.

Co-authored-by: Michael Lahargou <michaellahargou@gmail.com>

mlahargou reviewed (approved):
CR 📱
